File: 7.二叉树/5.二叉树的层序遍历/104.二叉树的最大深度.cpp
/*
 * @lc app=leetcode.cn id=104 lang=cpp
 *
 * [104] 二叉树的最大深度
 */
#include <bits/stdc++.h>
#include <ranges>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        TreeNode* q[10005];
        int front = 0, rear = 0, end = 0, level = 0;
        if (root) q[rear ++] = root, end ++;
        while (front < rear)
        {
            TreeNode* tmp = q[front ++];
            if (tmp->left) q[rear ++] = tmp->left;
            if (tmp->right) q[rear ++] = tmp->right;
            if (front == end)
            {
                end = rear;
                level ++;
            }
        }
        return level;
    }
};
// @lc code=end

File: HOT100/8.二叉树/104.二叉树的最大深度.py
#
# @lc app=leetcode.cn id=104 lang=python3
#
# [104] 二叉树的最大深度
#
from typing import Optional
from collections import deque
class TreeNode:
    def __init__(self, _val, _left, _right):
        self.val = _val
        self.left = _left
        self.right = _right
# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root: return 0
        l = self.maxDepth(root.left)
        r = self.maxDepth(root.right)
        return 1 + max(l, r)
# @lc code=end

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        q = deque()
        level = 0
        if root:
            q.append(root)
        while q:
            level += 1
            n = len(q)
            for _ in range(n):
                tmp = q.popleft()
                if tmp.left: q.append(tmp.left)
                if tmp.right: q.append(tmp.right)
        return level